### DISCUSSION TITLE
More trouble using Alpine Linux on Docker

### DISCUSSION THREAD

**Discussion by jdholtz** [seleniumbase/SeleniumBase#2135] [Q&A] [answered]
Hey. I am getting the following error about a version mismatch while using the `python:3.11-alpine` docker image with the latest version of SeleniumBase. This version mismatch is actually causing me to be detected.

```
The chromedriver version (117.0.5938.62) detected in PATH at /usr/bin/chromedriver might not be compatible with
the detected chrome version (117.0.5938.92); currently, chromedriver 117.0.5938.92 is recommended for chrome
117.*, so it is advised to delete the driver in PATH and retry
```

Here is the Docker image I am using (just do `docker build -t <name> .`, run it with `docker run -d <name>` and exec into it with `docker exec -it <name> sh`):
```dockerfile
FROM python:3.11-alpine

WORKDIR /app

RUN apk add --update --no-cache chromium chromium-chromedriver

RUN pip3 install --upgrade pip && pip3 install --no-cache-dir seleniumbase

ENTRYPOINT ["sleep", "infinity"]
```

Once in the Docker image, run `python -c "from seleniumbase import Driver; Driver()"` to see the issue.

I did not use `sbase` to download the driver because I was running into a similar issue to the one I described [here](https://github.com/seleniumbase/SeleniumBase/discussions/2111) when I used `sbase`.

How can I ensure that the SeleniumBase driver version matches the browser version so I don't get the warning above? Let me know if I can provide any more information to reproduce the issue.

**Comment by mdmintz** [comment, marked as answer]
By default, seleniumbase only downloads a new driver if the major version doesn't match (eg. 117). In the next release, SeleniumBase will let you set an exact version (https://github.com/seleniumbase/SeleniumBase/issues/2137). Also, based on your warning message, either an error occurred or seleniumbase wasn't used at all, because that warning message shouldn't appear when running tests via seleniumbase. SeleniumBase lets you have mismatched driver/browser versions (Eg. `driver_version="112"`, even if your Chrome version is 117.)

Based on your comments from https://github.com/seleniumbase/SeleniumBase/discussions/2111, it seems like you already figured out how to make it work:

<img width="936" alt="Screenshot 2023-09-24 at 11 26 11 AM" src="https://github.com/seleniumbase/SeleniumBase/assets/6788579/b51255b0-d4cd-4f1f-8b2b-a14bdf498e09">

Also note that `Driver()` doesn't use UC Mode by default.

Make sure to use UC Mode if you want `chromedriver` to remain undetected: `--uc` / `uc=True`.

Example test: https://github.com/seleniumbase/SeleniumBase/blob/master/examples/raw_uc_mode.py

**Comment by jdholtz** [comment]
> Based on your comments from https://github.com/seleniumbase/SeleniumBase/discussions/2111, it seems like you already figured out how to make it work

That did fix the issue I was receiving in that discussion, but I unfortunately am receiving the browser mismatch with the fix I was referring to.

Instead of selecting the exact version to match, is there a way for SeleniumBase to select the same version that the browser uses so there is no mismatch? Is that what `--driver-version="keep"` is? I tried that, but it may not have worked due to #2136. 

This would help in not needing to get the browser version beforehand (as it is in a Docker image, so it will change every time it is rebuilt, if a new version has been released for Alpine Linux), but instead letting SB do the work.

**Reply by mdmintz** [reply]
I created https://github.com/seleniumbase/SeleniumBase/issues/2139 for it. (`--driver-version="browser"` / `driver_version="browser"`). Will be in the next release.

**Reply by jdholtz** [reply]
Awesome. Thanks!

**Reply by mdmintz** [reply]
New release is ready: `4.18.9` - https://github.com/seleniumbase/SeleniumBase/releases/tag/v4.18.9

**Reply by jdholtz** [reply]
Sweet! However, I did have an issue with using `headless=True` and `driver_version="browser"`, as headless is not respected while using the browser version. I tried `headless2=True`, but unfortunately I am being detected. Also, in the Docker image, I noticed that using `driver_version="browser"` would download a new Chromedriver (basically print `Warning: uc_driver not found. Getting it now:...`), but neglecting that parameter would not (even though the Chromedriver and Chromium browser are the same version already).

An interesting thing to note is that upgrading to Selenium 4.13.0 gets rid of the warning about the version mismatch (the very first thing in this post). However, it still resulted in me being detected.

**Reply by mdmintz** [reply]
`uc_driver` doesn't support ARM versions. On an ARM machine, such as an M1 Mac, it will will download the Intel version and then convert that driver to an undetectable one, which works. Due to such differences, the `uc_driver` is acquired from a downloaded `chromedriver`, even if there may already be a `chromedriver` downloaded from a non-UC run. If `uc_driver` is already there, it won't download it again unless the major version isn't a match. If SeleniumBase didn't encounter any errors, then warnings such as `might not be compatible with the detected chrome version` should not be happening because that is hidden with UC Mode and standard runs. If you saw it, it means there was an error somewhere else, (and it didn't use UC Mode), so I would need more debug logs from that to see what happened. Also, I'll be upgrading seleniumbase to selenium `4.13.0` sometime today, so that message will get hidden, but that message is a clue that something else is up.

**Reply by jdholtz** [reply]
Sounds good. Just to note, I am not using ARM for any issues I have run into on this discussion.

**Comment by jdholtz** [comment]
Hey @mdmintz, I believe this could be solved if I can get the `sbase get chromedriver` command to work. With the Docker image I provided above (as a Dockerfile), I get the following output that it successfully installed:
```

*** chromedriver to download = 117.0.5938.92 (Latest Stable) 

Downloading chromedriver-linux64.zip from:
https://edgedl.me.gvt1.com/edgedl/chrome/chrome-for-testing/117.0.5938.92/linux64/chromedriver-linux64.zip ...
Download Complete!

Extracting ['chromedriver'] from chromedriver-linux64.zip ...
Unzip Complete!

The file [uc_driver] was saved to:
/usr/local/lib/python3.11/site-packages/seleniumbase/drivers/uc_driver

Making [uc_driver 117.0.5938.92] executable ...
[uc_driver 117.0.5938.92] is now ready for use!
```

However, I get the following error when trying to execute the driver (through the CLI, not using Python).
```sh
# /usr/local/lib/python3.11/site-packages/seleniumbase/drivers/uc_driver  --version
sh: /usr/local/lib/python3.11/site-packages/seleniumbase/drivers/uc_driver: not found
```

I tried this on the regular Ubuntu image (not Alpine) and it successfully displays the version. Therefore, I may be missing a package that helps download the driver correctly. Do you know of anything that could be causing this issue and can you reproduce it yourself?

There aren't any tracebacks as this is all done through the command line, and I am using SB 4.18.9. Thank you very much for your help on this issue.

**Reply by mdmintz** [reply]
Either the `uc_driver` was deleted after you downloaded it, or perhaps an issue with regular undetected-chromedriver on Alpine Linux Docker: https://github.com/search?q=repo%3Aultrafunkamsterdam%2Fundetected-chromedriver+alpine+docker&type=issues

**Reply by jdholtz** [reply]
The issue ended up being missing a dependency (ld-linux-x86-64.so.2). I downloaded the `gcompat` package and then the file downloaded using sbase was able to be executed successfully. Thanks!
